### PR TITLE
Refactor tag assertion to FatalError check & Fix unexpected increase of tag (fork)

### PR DESCRIPTION
**Pull request by soywod**
This is just a fork of https://github.com/jonhoo/rust-imap/pull/231 so it can be finished and merged.

- - -
This change is [Reviewable](https://reviewable.io/reviews/jonhoo/rust-imap/284)

### CODE REVIEW

**Comment by soywod** [comment]
One [test fails](https://github.com/jonhoo/rust-imap/actions/runs/8408073328/job/23023789923?pr=284) but it is not related to this PR.

File: src/error.rs [hunk 1]
```diff
@@ -296,6 +301,36 @@ impl StdError for ValidateError {
...
+}
+
+impl StdError for TagMismatch {
+    fn description(&self) -> &str {
+        "Tag is corrupted, session is in an inconsistent state"
+    }
+
+    fn cause(&self) -> Option<&dyn StdError> {
+        None
+    }
```
Comment (jonhoo) [inline src/error.rs:331]: You can remove both of these impls — their default impl is fine.

File: src/error.rs [hunk 2]
```diff
@@ -296,6 +301,36 @@ impl StdError for ValidateError {
...
 }
 
+/// Tag was corrupted during session.
+#[derive(Debug)]
+#[non_exhaustive]
+pub struct TagMismatch {
+    /// Expected tag number
+    pub(crate) expect: u32,
+    /// Actual tag number, 0 if parse failed
+    pub(crate) actual: u32,
```
Comment (jonhoo) [inline src/error.rs:311]: Let's make this a `Result<u32, Vec<u8>>` instead so we can directly present the unparseable tag if present.

File: src/error.rs [hunk 3]
```diff
@@ -296,6 +301,36 @@ impl StdError for ValidateError {
     }
 }
 
+/// Tag was corrupted during session.
+#[derive(Debug)]
```
Comment (jonhoo) [inline src/error.rs:305]:
```suggestion
#[derive(Debug, Clone, PartialEq, Eq)]
```

File: src/error.rs [hunk 4]
```diff
@@ -105,6 +105,8 @@ pub enum Error {
     /// In response to a STATUS command, the server sent OK without actually sending any STATUS
     /// responses first.
     MissingStatusResponse,
+    /// Tag mismatch between client and server. New session must be created.
```
Comment (jonhoo) [inline src/error.rs:108]:
```suggestion
    /// The server responded with a different command tag than the one we just sent.
    ///
    /// A new session must generally be established to recover from this.
```

File: src/error.rs [hunk 5]
```diff
@@ -296,6 +301,36 @@ impl StdError for ValidateError {
     }
 }
 
+/// Tag was corrupted during session.
```
Comment (jonhoo) [inline src/error.rs:304]:
```suggestion
/// The server responded with a different command tag than last one we sent.
```

File: src/client.rs [hunk 6]
```diff
@@ -155,6 +157,13 @@ pub struct Client<T: Read + Write> {
     conn: Connection<T>,
 }
 
+impl<T: Read + Write> Client<T> {
```
Comment (jonhoo) [inline src/client.rs:160]: Let's move this to be on `Connection` instead.

File: src/client.rs [hunk 7]
```diff
@@ -155,6 +157,13 @@ pub struct Client<T: Read + Write> {
...
+    /// Manually increment the current tag.
```
Comment (jonhoo) [inline src/client.rs:161]: Could you expand the docs a bit more here to indicate why someone may want to use this method?

**Review by jonhoo — CHANGES_REQUESTED** [review]
Thanks for picking this up!

**Comment by soywod** [comment]
Ready for a second review.

**Review by jonhoo — APPROVED** [review]
Thanks! I want to tweak the docs a little more, but can do that locally. Let's get this merged :tada:

**Comment by soywod** [comment]
> Let's get this merged

Awesome, thank you! Do you think it could be possible to release a new
alpha version? I need this fix to release a new version of mine.

**Comment by jonhoo** [comment]
Yep, it's coming shortly! I'm preparing it over here: https://github.com/jonhoo/rust-imap/pull/285

The specific doc changes I made are these: https://github.com/jonhoo/rust-imap/pull/285/commits/ed433ec5745a3dfb1aaecc22bbbe5d834d5c33dd

**Comment by jonhoo** [comment]
Published as 3.0.0-alpha.13 :tada:

**Pull request merged**